numba/tests/test_dyn_array.py

```diff
@@ -0,0 +1,362 @@
+        def pyfunc(y):
+            return y
+
+        arr = np.ones(4, dtype=np.float32)
+
+        cfunc = nrtjit(pyfunc)
+        expected = cfunc(arr)
+        got = pyfunc(arr)
+
+        np.testing.assert_equal(expected, arr)
+        np.testing.assert_equal(expected, got)
+        self.assertIs(expected, arr)
+        self.assertIs(expected, got)
+
+    def test_array_pass_through_sliced(self):
+        def pyfunc(y):
+            return y[y.size // 2:]
+
+        arr = np.ones(4, dtype=np.float32)
+
+        initrefct = sys.getrefcount(arr)
+
+        cfunc = nrtjit(pyfunc)
+        got = cfunc(arr)
+        self.assertEqual(initrefct + 1, sys.getrefcount(arr))
+        expected = pyfunc(arr)
+        self.assertEqual(initrefct + 2, sys.getrefcount(arr))
+
+        np.testing.assert_equal(expected, arr[arr.size // 2])
+        np.testing.assert_equal(expected, got)
+
+        del expected
+        self.assertEqual(initrefct + 1, sys.getrefcount(arr))
+        del got
+        self.assertEqual(initrefct, sys.getrefcount(arr))
+
+    def test_ufunc_with_allocated_output(self):
+
+        def pyfunc(a, b):
+            out = np.empty(a.shape)
+            np.add(a, b, out)
+            return out
+
+        cfunc = nrtjit(pyfunc)
+
+        # 1D case
+        arr_a = np.random.random(10)
+        arr_b = np.random.random(10)
+
+        np.testing.assert_equal(pyfunc(arr_a, arr_b),
+                                cfunc(arr_a, arr_b))
+
+        # 2D case
+        arr_a = np.random.random(10).reshape(2, 5)
+        arr_b = np.random.random(10).reshape(2, 5)
+
+        np.testing.assert_equal(pyfunc(arr_a, arr_b),
+                                cfunc(arr_a, arr_b))
+
+        # 3D case
+        arr_a = np.random.random(70).reshape(2, 5, 7)
+        arr_b = np.random.random(70).reshape(2, 5, 7)
+
+        np.testing.assert_equal(pyfunc(arr_a, arr_b),
+                                cfunc(arr_a, arr_b))
+
+    def test_multithread(self):
+
+        def pyfunc(inp):
+            out = np.empty(inp.size)
+            tmp = 0
+            for i in range(out.size):
+                out[i] = tmp
+                tmp = inp[i]
+            return out
+
+        cfunc = nrtjit(pyfunc)
+        size = 10**5
+        arr = np.random.randint(0, 1, size)
+
+        np.testing.assert_equal(pyfunc(arr), cfunc(arr))
+
+        workers = []
+        inputs = []
+        outputs = []
+
+        # Make wrapper to store the output
+        def wrapped(inp, out):
+            out[:] = cfunc(inp)
+
+        # Create worker threads
+        for i in range(10):
+            arr = np.random.randint(0, 1, size)
+            out = np.empty_like(arr)
+            thread = threading.Thread(target=wrapped,
+                                      args=(arr, out),
+                                      name="worker{0}".format(i))
+            workers.append(thread)
+            inputs.append(arr)
+            outputs.append(out)
+
+        # Launch worker threads
```

Comment:
I'm not sure this tests much. The threads will be serialized, since you didn't add `nogil=True`. Furthermore, they all operate on distinct arrays, so there's not much chance for contention (is there?).